### ISSUE TITLE
Middleware stacking and recursive initialization

### ISSUE THREAD

**Issue by Kludex** [Kludex/starlette#2002]
### Discussed in https://github.com/encode/starlette/discussions/1161

<div type='discussions-op-text'>

<sup>Originally posted by **euri10** April  1, 2021</sup>
long story short I was writing a logging middleware whose handlers are initialized in the `__init__` of the middleware and ended up with multiple logs.

it seems to me the `__init__` of the middlewarea are recursively called  because of the `build_middleware_stack` method and I'm not sure if this is intended or a bug.

here's a test case that shows what I think is an issue:

```
class SimpleInitializableMiddleware:
    def __init__(self, app, counter: int):
        self.app = app
        self.counter = counter
        print(self.counter)


def test_middleware_stack_init():
    app = Starlette()
    app.add_middleware(SimpleInitializableMiddleware, counter=1)
    app.add_middleware(SimpleInitializableMiddleware, counter=2)
    app.add_middleware(SimpleInitializableMiddleware, counter=3)
    app.add_middleware(SimpleInitializableMiddleware, counter=4)
```

I would expect to see `1 2 3 4` as the output.
however it prints `1 1 2 1 2 3 1 2 3 4` because for each add_middlware call it goes down the stack and reinitialize it.

what you think @tomchristie ?</div>

**Comment by Kludex** [comment]
A possible solution would be to deprecate, and further remove `add_middleware()`.

**Comment by Kludex** [comment]
- We are going with https://github.com/encode/starlette/pull/2017. Check more about it over there.

This will be available on Starlette 0.24.0.
